File: java/src/main/java/gov/doc/ntia/sigmf/Acquisition.java
package gov.doc.ntia.sigmf;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import gov.doc.ntia.sigmf.ext.global.scos.Action;
import gov.doc.ntia.sigmf.ext.global.scos.ScheduleEntry;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.nio.ByteBuffer;

public class Acquisition implements Serializable {


    private static final long serialVersionUID = 1L;

    @JsonProperty
    protected MetaDoc metaDoc;

    @JsonProperty
    protected transient ByteBuffer data;

    public Acquisition(){}


    public MetaDoc getMetaDoc() {
        return metaDoc;
    }

    public void setMetaDoc(MetaDoc metaDoc) {
        this.metaDoc = metaDoc;
    }

    public ByteBuffer getData() {
        return data;
    }

    public void setData(ByteBuffer data) {
        this.data = data;
    }


    @Override
    public String toString(){
        ObjectMapper mapper = new ObjectMapper();
        String message=null;
        try {
            return mapper.writeValueAsString(this);
        } catch (JsonProcessingException e) {
            e.printStackTrace();
            message = e.getMessage();
        }
        return message;
    }


    private void writeObject(ObjectOutputStream oos)
            throws IOException {
        oos.defaultWriteObject();
        oos.writeObject(metaDoc);
        oos.writeObject(this.data.array());
    }

    private void readObject(ObjectInputStream ois)
            throws ClassNotFoundException, IOException {
        ois.defaultReadObject();
        metaDoc = (MetaDoc) ois.readObject();
        byte[] bytes = (byte[])ois.readObject();
        data = ByteBuffer.wrap(bytes);
    }

    public String getName(){
        String name = getMetaDoc().getGlobal().getSensor().getId() + "_" +
                getScheduleId()+ "_" +
                getTaskId();

        return name;
    }

    public String getActionName(){
        Action action = metaDoc.getGlobal().getAction();
        if(action == null || action.getName() == null){
            return "unknown";
        }
        else{
            return action.getName();
        }
    }


    private String getScheduleId(){
        ScheduleEntry scheduleEntry = metaDoc.getGlobal().getSchedule();
        String missingScheduleInfo = "unknown";
        if(scheduleEntry == null){
            return  missingScheduleInfo;
        }
        else if(scheduleEntry.getId() == null){
            return missingScheduleInfo;
        }
        else{
            return scheduleEntry.getId();
        }
    }

    private  String getTaskId(){
        String taskId = metaDoc.getGlobal().getTaskId().toString();
        if(taskId == null){
            taskId = "unknown";
        }
        return taskId;
    }

}

File: C#/sigmf/sigmf/Acquisition.cs
﻿
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.IO;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;

namespace sigmf
{
    [Serializable()]
    public class Acquisition
    {
        private static readonly long serialVersionUID = 1L;

        [JsonProperty]
        public MetaDoc metaDoc;

        [JsonProperty]
        [JsonIgnore]
        public MemoryStream data;

        public Acquisition() {
            metaDoc = new MetaDoc();
            data = new MemoryStream();
        }

        public MetaDoc getMetaDoc() {
            return metaDoc;
        }
        public void setMetaDoc(MetaDoc metaDoc)
        {
            this.metaDoc = metaDoc;
        }
        public MemoryStream getData()
        {
            return data;
        }
        public void setData(MemoryStream data)
        {
            this.data = data;
        }

        public MemoryStream writeObject(MetaDoc meta)
        {
            MemoryStream stream = new MemoryStream();
                
            IFormatter formatter = new BinaryFormatter();
            formatter.Serialize(stream, meta);
            return stream;
            
        }
        public MetaDoc readObject(MemoryStream stream)
        {
            IFormatter formatter = new BinaryFormatter();
            stream.Seek(0, SeekOrigin.Begin);
            MetaDoc meta = (MetaDoc)formatter.Deserialize(stream);
            return meta;
        }

        public string toString()
        {
            var acq = JsonConvert.SerializeObject(this);
            return acq;
        }
        public string GetName() {
            string name = getMetaDoc().getGlobal().Sensor.Id + "_" + getScheduleId() + "_" + getTaskId();
            return name;
        }
        public string getActionName()
        {
            Action action = metaDoc.getGlobal().Action;
            if(action == null || action.Name == null)
            {
                return "unknown";
            }
            else
            {
                return action.Name;
            }
        }
        public string getScheduleId()
        {
            ScheduleEntry scheduleEntry = metaDoc.getGlobal().Schedule;
            string missingScheduleInfo = "unknown";
            if(scheduleEntry == null)
            {
                return missingScheduleInfo;
            }
            else if(scheduleEntry.Id == null)
            {
                return missingScheduleInfo;
            }
            else
            {
                return scheduleEntry.Id;
            }
        }
        public string getTaskId()
        {
            string taskId = metaDoc.getGlobal().TaskId.ToString();
            if(taskId == null)
            {
                taskId = "unknown";
            }
            return taskId;
        }

    }
}